runtime/limits_integ_test.go

```diff
@@ -0,0 +1,174 @@
+	"testing"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/oci"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+type commandResult struct {
+	stdout   string
+	stderr   string
+	exitCode uint32
+}
+
+func runTask(ctx context.Context, c containerd.Container) (*commandResult, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	task, err := c.NewTask(ctx, cio.NewCreator(cio.WithStreams(nil, &stdout, &stderr)))
+	if err != nil {
+		return nil, err
+	}
+
+	exitCh, err := task.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = task.Start(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case exitStatus := <-exitCh:
+		if err := exitStatus.Error(); err != nil {
+			return nil, err
+		}
+
+		_, err := task.Delete(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return &commandResult{
+			stdout:   stdout.String(),
+			stderr:   stderr.String(),
+			exitCode: exitStatus.ExitCode(),
+		}, nil
+	case <-ctx.Done():
+		return nil, errors.New("context cancelled")
+	}
+}
+
+func TestDiskLimit_Isolated(t *testing.T) {
+	internal.RequiresIsolation(t)
+
+	assert := assert.New(t)
+	require := require.New(t)
+
+	ctx := namespaces.WithNamespace(context.Background(), "default")
+
+	client, err := containerd.New(containerdSockPath, containerd.WithDefaultRuntime(firecrackerRuntime))
+	require.NoError(err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	image, err := alpineImage(ctx, client, naiveSnapshotterName)
+	require.NoError(err, "failed to get alpine image")
+
+	sh := containerd.WithNewSpec(oci.WithProcessArgs("sh", "/var/firecracker-containerd-test/scripts/diskhog.sh"),
+		oci.WithDefaultPathEnv,
+		oci.WithMounts([]specs.Mount{
+			{
+				Type:        "bind",
+				Options:     []string{"bind"},
+				Destination: "/var/firecracker-containerd-test/scripts",
+				Source:      "/var/firecracker-containerd-test/scripts",
+			},
+		}))
+
+	container, err := client.NewContainer(ctx,
+		"container",
+		containerd.WithSnapshotter(naiveSnapshotterName),
+		containerd.WithNewSnapshot("snapshot", image),
+		sh,
+	)
+	defer func() {
+		err = container.Delete(ctx, containerd.WithSnapshotCleanup)
+		require.NoError(err, "failed to delete a container")
+	}()
+
+	result, err := runTask(ctx, container)
+	require.NoError(err, "failed to create a container")
+
+	assert.Equal(uint32(1), result.exitCode, "exitCode")
+
+	assert.Equal(`Filesystem                Size      Used Available Use% Mounted on
+/dev/vdb                975.9M      8.4M    900.3M   1% /
+`, result.stdout, "stdout")
+
+	// TODO: Doesn't it have only 900MB? The written records are higher than expected.
```

Comment:
Where did the 975.9M disk come from?  It doesn't look like that's defined as part of the test, but rather an implementation detail of the naive snapshotter.  Is that correct?

Reply:
Yes. Coming from https://github.com/firecracker-microvm/firecracker-containerd/blob/6218f7e43acfa439b5ac299677aa7a1458796470/snapshotter/naive/naive.go#L38, I think.

Let me check whether we have a way to have an explicit limit or not.

Reply:
Yes. With devmapper, the size is much much smaller.

```
--- FAIL: TestDiskLimit_Isolated (6.98s)
    limits_integ_test.go:118:
                Error Trace:    limits_integ_test.go:118
                Error:          Not equal:
                                expected: "Filesystem                Size      Used Available Use% Mounted on\n/dev/vdb                975.9M      8.4M    900.3M   1% /\n"
                                actual  : "Filesystem                Size      Used Available Use% Mounted on\n/dev/vdb                120.0M      7.0M    104.0M   6% /\n"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,3 +1,3 @@
                                 Filesystem                Size      Used Available Use% Mounted on
                                -/dev/vdb                975.9M      8.4M    900.3M   1% /
                                +/dev/vdb                120.0M      7.0M    104.0M   6% /

                Test:           TestDiskLimit_Isolated
                Messages:       stdout
    limits_integ_test.go:123:
                Error Trace:    limits_integ_test.go:123
                Error:          Not equal:
                                expected: "952+0 records in\n951+0 records out\n"
                                actual  : "111+0 records in\n110+0 records out\n"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,3 +1,3 @@
                                -952+0 records in
                                -951+0 records out
                                +111+0 records in
                                +110+0 records out

                Test:           TestDiskLimit_Isolated
                Messages:       stderr
FAIL
```

Reply:
Because I'm specifying https://github.com/firecracker-microvm/firecracker-containerd/blob/38742eeb4fb5d549b31727a57b8c79e88b204b59/tools/docker/entrypoint.sh#L29 🤦‍♂

Reply:
Commented about where the size is coming from. We can make naive snapshotter more configurable, but the snapshotter is not for production. So, not so sure the value it brings vs. complexity.